controller/rancher/certificate_fetcher.go

```
@@ -0,0 +1,317 @@
+
+	toUpdate := make(map[string]interface{})
+	toUpdate["publicEndpoints"] = eps
+	logrus.Infof("Updating Rancher LB [%s] in stack [%s] with the new public endpoints [%v] ", lbSvc.Name, lbSvc.StackName, eps)
+	if _, err := fetcher.Client.LoadBalancerService.Update(&lb, toUpdate); err != nil {
+		return fmt.Errorf("Failed to update Rancher LB [%s] in stack [%s]. Error: %#v", lbSvc.Name, lbSvc.StackName, err)
+	}
+	return nil
+}
+
+/* Cache the certs present under CertDir and DefaultCertDir
+   When the Cache is null, read from the dir and init the cache. When cache is not empty read from the cache.
+   Clear cache when the directories change
+   Start a Watcher thread when initialising the cache. For every certName, add to watcher.
+   When watcher receives event, read that cert and update cache.
+   Add a watcher for CertDir/DefaultCertDir also - upon events on these dirs clear the cache and enque.*/
+
+func (fetcher *RCertificateFetcher) ReadAllCertificatesFromDir(certDir string) ([]*config.Certificate, error) {
+	if _, err := os.Stat(certDir); os.IsNotExist(err) {
+		return nil, fmt.Errorf("Certificate Directory does not exist %v", certDir)
+	}
+	certs := []*config.Certificate{}
+
+	fetcher.mu.RLock()
+	currentCertD := fetcher.CertDir
+	fetcher.mu.RUnlock()
+
+	if currentCertD != "" && currentCertD == certDir && fetcher.CertsCache != nil && len(fetcher.CertsCache) > 0 {
+		//read from cache if certDir matches and certs are cached.
+		logrus.Debug("Reading certs from cache")
+	} else {
+		// init/re-init cache
+		fetcher.mu.Lock()
+		fetcher.removeWatches()
+
+		fetcher.CertsCache = make(map[string]*config.Certificate)
+		logrus.Debug("Reading certs from directory and starting watchers")
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(certDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.CertDir = ""
+			fetcher.CertsCache = nil
+			logrus.Errorf("Error %v reading certs from directory %v", err, certDir)
```

Should we return an error here? or the intent is to read the cert from the directory ignoring the error?